### ISSUE TITLE
How to contribute

### ISSUE THREAD

**Issue by francisco-dlp** [hyperspy/hyperspy#1170]
The following is a suggestion on how to contribute to HyperSpy to be added to the Dev Guide after discussion here:

Following #904  and #1060 there are three new _permanent_ branches:
- RELEASE_next_major
- RELEASE_next_minor
- RELEASE_next_patch

~~The master is the nickname of the (non-existent) RELEASE_next_major branch.~~

Therefore, when working on new pull requests _branch_ and _pull request to_:
- RELEASE_next_major **only** if what you're working on requires changing the API
- RELEASE_next_minor if you're planning to add a new feature or enhance an existing one.
- RELEASE_next_patch if you are fixing a bug in the code or improving the docs.

The default branch is RELEASE_next_minor.

Any new PR, if considered for inclusion in HyperSpy, will be labelled one of RELEASE_next_minor, RELEASE_next_major, RELEASE_next_patch. Once the PR is merged, it'll be added to the milestone of the following release of its kind i.e. next major increment for a major, next minor increment for a minor and next patch increment for a patch PR. Once a milestone is closed all RELEASE labels of the associated PRs must be removed.

How to manage new PRs - Summary for developers:
- [ ] Verify that the author has branched from the correct RELEASE branch
- [ ] Add RELEASE label
- [ ] When closing release milestone remove RELEASE label

**Comment by dnjohnstone** [comment]
I think this is all good, but should be added to the Dev Guide ASAP so that people can get there branches in order before everything gets massively out of sync.

On a practical note - it seems to me that if I have a branch adding a new feature that is currently based on the master then all I need to do at this point is to merge it with RELEASE_next_minor and then keep doing that (crucially not with master) . Anyone who has big branch changes locally will appreicate getting that right the first time as it's a real pain to unpick after (yes I did it, was too used to merging master).

Secondly - do you want me to knock this into words for the Dev Guide? I know that words come quite quickly to me in English.

**Comment by francisco-dlp** [comment]
Yes, there is no difference between RELEASE_next_minor and master at present, so that approach should work... but not for a long time.

If you're happy to add this to the Dev Guide, please do.

**Comment by francisco-dlp** [comment]
To avoid confusion, I have just renamed master to RELEASE_next_major. Also, RELEASE_next_minor is now the default branch.

**Comment by thomasaarholt** [comment]
I'm a big fan of this. There might be a bit of confusion with the term "master" now, but that should be ok. I really like that the branches have been cleared up - I remember being really confused when I first looked through that.

**Comment by francisco-dlp** [comment]
@thomasaarholt, I am glad that you like the new structure. It seems such an obvious choice that I wonder how it didn't occur to us before.

There is no longer confusion with master because we have deleted the master branch. The default branch is now RELEASE_next_minor.

**Comment by tjof2** [comment]
Unless it's in an odd location, it looks like the details of the different branches are still missing from the dev guide @francisco-dlp

**Comment by ericpre** [comment]
This is now in http://hyperspy.org/hyperspy-doc/current/dev_guide/git.html#semantic-versioning-and-hyperspy-main-branches.
